frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+//! the error case (Goal 6).  
+//! Weight limits must be checked by the migration itself. The pallet provides a [`WeightMeter`] for
+//! that cause. The pallet may return [`SteppedMigrationError::InsufficientWeight`] at any point.
+//! The pallet will react to this with a case decision: if that migration was exclusively execute in
+//! this block, and therefore got the maximal amount of weight possible, the pallet becomes `Stuck`.
+//! Otherwise one re-attempt is done with the same logic in the next block (Goal 3).  
+//! Progress of the pallet is guaranteed by providing once: a timeout for each migration via
+//! [`SteppedMigration::max_steps`]. The pallet **ONLY** guarantees progress if this is set to
+//! sensible limits (Goal 7).
+//!
+//! ### Scenario: Governance cleanup
+//!
+//! Every now and then, governance can make use of the `clear_historic` call. This ensures that no
+//! old migrations pile up in the `Historic` set. This can probably be done very rarely, since the
+//! storage should not grow quickly and the lookup weight does not suffer much.
+//!
+//! ### Scenario: Successful upgrade
+//!
+//! The standard procedure for a successful runtime upgrade can look like this:
+//! 1. Migrations are configured in the `Migrations` config item. All migrations expose `max_steps`,
+//! error tolerance, check their weight bounds and have a unique identifier.  
+//! 2. The runtime upgrade is enacted. Events `UpgradeStarted` is followed by lots of
+//! `MigrationAdvanced`. Finally an `UpgradeCompleted` is emitted.  
+//! 3. Cleanup as described on the governance scenario can happen at any time.
+//!
+//! ### Advice: Failed upgrades
+//!
+//! Failed upgrade cannot automatically be handled but requires governance intervention. Set up
+//! monitoring for event `UpgradeFailed` to act in this case. Hook [`UpgradeStatusHandler::failed`]
+//! should be setup in a way that it allows governance to act, but still prevent other transactions
+//! from interacting with the inconsistent storage state. Note that this is paramount, since the
+//! inconsistent state might contain faulty balance amount or similar that could cause great harm if
+//! user transactions don't remain suspended. One way to implement this would be to use the
+//! `SafeMode` or `TxPause` pallets that can prevent most user interactions but still allow a
+//! whitelisted set of governance calls.
+//!
+//! ### Remark: Failed migrations
+//!
+//! Failed migrations are not added to the `Historic` blacklist. This means that an erroneous
+//! migration must be removed of fixed manually. This already applies - even before taking the
+//! historic set into account.
+//!
+//! ### Remark: Transactional processing
+//!
+//! You can see the transactional semantics for migrational steps as mostly useless, since in the
+//! stuck case the state is already messed up. This just prevents it from getting messed up even
+//! more, but does not prevent it in the first place.
```

Comment:
```suggestion
//! You can see the transactional semantics for migrational steps as mostly useless, since in the
//! stuck case the state is already messed up. This just prevents it from becoming even more messed up
//! , but doesn't prevent it in the first place.
```